File: algorithm/437. Path Sum III/437.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        if not root:
            return 0
        return self.helper(root,sum)+self.pathSum(root.left,sum)+self.pathSum(root.right,sum)
        
    def helper(self, root, sum):
        if not root:
            return 0
        return int(root.val==sum)+self.helper(root.left,sum-root.val)+self.helper(root.right,sum-root.val)
            
        

File: algorithm/437. Path Sum III/437.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int sum) {
        if (!root) return 0;
        return helper(root,sum,sum)+pathSum(root->left,sum)+pathSum(root->right,sum) ;
    }
    
    int helper(TreeNode* root, int sum,int nowsum) {
        if(!root) return 0;
        if(root->val==nowsum) return 1+ helper(root->left,sum, nowsum-root->val) + helper(root->right,sum, nowsum-root->val);
        return helper(root->left,sum, nowsum-root->val) + helper(root->right,sum, nowsum-root->val);
    }
};
